README
NAME
    Term::GnuScreen - Control GNU screen via perl

VERSION
    Version 0.03

SYNOPSIS
    Term::GnuScreen provides a simple interface to control a GNU screen
    session via its command line interface.

        use Term::GnuScreen;

        my $screen = Term::GnuScreen->new();
        $screen->windowlist;
        $screen->hardcopy('/tmp/my_hardcopy');

CONSTRUCTION
    session
        Sets the name of the screen session to send commands to. If you also
        set "create" to a true value, this will become the new name of your
        screen session. See *-S* option for screen for a further discussion
        of this argument.

    create
        If create is set to a true value, a new screen session is created
        and detached automatically. If you do not provide a session name,
        this module generates one by calling ""term_gnuscreen" . $$ .
        int(rand(10000))". Settings this value after object creation has no
        effect at the moment.

        The newly created session will not be terminated after programm
        execution.

    window
        Preselects a window to send a command via the a specific window.
        Defaults to 0. See *-p* option for screen for a further discussion
        of this argument.

    executable
        Return or set the screen binary to call. Defaults to the binary
        found by File::Which::which.

METHODS
    Term::GnuScreen implements all commands as stated in the texinfo
    document shipped with GNU screen. To call a commands, it's send via GNU
    screens -X paramter to the first running screen session and its current
    window. You can change session and window with the according object
    methods and construction paramters. Unless listed here, all remaining
    arguments are handled over to screen without further modification.

    The five commands bind, meta, chdir, exec and umask are prefixed with a
    *s* ( sbind, smeta, schdir, sexec and sumas ) to distinguish them from
    the built-ins with the same name.

  call_screen
    This command is the working horse of Term::GnuScreen. It simply builds
    the command line to call and execute it.

  send_command
    Calls call_screen with the *-X* and all supplied paramters. Most
    functions are implemented by this method.

  hardcopy
    Write a hardcopy of the current window to a temporary file and returns
    the filename unless the filename is supplied as first argument. If the
    supplied filename is not absolute, the file is written relative to
    "hardcopydir".

ERROR HANDLING
    Simple dies in case screen -X did not return with a return value of
    zero. Either $!, STDERR or STDOUT (which seems to be more helpful most
    times) are provided as error message for further investigation.

AUTHOR
    Mario Domgoergen, "<dom at math.uni-bonn.de>"

BUGS AND LIMITATIONS
    It seems not to be possible to question a specific screen session about
    its state, so this module basically just sends commands to a screen
    session without knowing if the command succeeded or was at least
    syntactically corrent.

    This module needs a lot more testing.

    Please report any bugs or feature requests to "bug-term-gnuscreen at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Term-GnuScreen>. I will
    be notified, and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Term::GnuScreen

    You can also look for information at:

    * RT: CPAN's request tracker
        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Term-GnuScreen>

    * AnnoCPAN: Annotated CPAN documentation
        <http://annocpan.org/dist/Term-GnuScreen>

    * CPAN Ratings
        <http://cpanratings.perl.org/d/Term-GnuScreen>

    * Search CPAN
        <http://search.cpan.org/dist/Term-GnuScreen>

ACKNOWLEDGEMENTS
    screen

COPYRIGHT & LICENSE
    Copyright 2009 Mario Domgoergen, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Term-GnuScreen

0.03	Mo 10. Aug 13:46:57 CEST 2009
	added github respository to META.yml
	added dependency for IPC::System::Simple as autodies relies on it when using :all

0.02	Mo 10. Aug 11:59:36 CEST 2009
	clarify documentation on the five built-ins
	added construction parameter 'create'
	window defaults to 0 (documentation changed accordingly)
	added check for defindness of session and window (especially a window named 0 is common)


0.01    So 9. Aug 23:03:15 CEST 2009
        First version, released on an unsuspecting world.



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name       => 'Term::GnuScreen',
    license           => 'perl',
    dist_author       => 'Mario Domgoergen <dom@math.uni-bonn.de>',
    dist_version_from => 'lib/Term/GnuScreen.pm',
    requires          => {
        Moose               => 0,
        'Sub::Install'      => 0,
        'File::Temp'        => 0,
        autodie             => 0,
        'File::Which'       => 0,
        'IO::CaptureOutput' => 0,
	'IPC::System::Simple' => 0,
    },
    build_requires     => { 'Test::More' => 0, },
    add_to_cleanup     => ['Term-GnuScreen-*'],
    create_makefile_pl => 'traditional',
    create_readme      => 1,
    meta_merge         => {
        resources => { repository => 'http://github.com/mdom/term-gnuscreen', },
    }
);

$builder->create_build_script();


lib/Term/GnuScreen.pm
package Term::GnuScreen;

use Moose;
use Sub::Install qw(install_sub);
use File::Temp qw(tmpnam);
use autodie qw(:all);
use File::Which;
use IO::CaptureOutput qw(capture);

our $VERSION = '0.03';

BEGIN {

	my @commands = ( qw( acladd aclchg acldel aclgrp aclumask activity addacl allpartial
	altscreen at attrcolor autodetach autonuke backtick bce bell_msg 
	bindkey blanker blankerprg break breaktype bufferfile c1 caption chacl
	charset clear colon command compacthist console copy copy_reg
	crlf debug defautonuke defbce defbreaktype defc1 defcharset defencoding
	defescape defflow defgr defhstatus deflog deflogin defmode defmonitor
	defnonblock defobuflimit defscrollback defshell defsilence defslowpaste
	defutf8 defwrap defwritelock defzombie detach digraph dinfo displays
	dumptermcap echo encoding escape eval fit flow focus gr 
	hardcopy_append hardcopydir hardstatus height help history hstatus idle
	ignorecase info ins_reg kill lastmsg license lockscreen log logfile login
	logtstamp mapdefault mapnotnext maptimeout markkeys maxwin monitor
	msgminwait msgwait multiuser nethack next nonblock number obuflimit only
	other partial password paste pastefont pow_break pow_detach pow_detach_msg
	prev printcmd process quit readbuf readreg redisplay register remove
	removebuf reset resize screen scrollback select sessionname setenv setsid
	shell shelltitle silence silencewait sleep slowpaste source sorendition
	split startup_message stuff su suspend term termcap terminfo termcapinfo
	time title unsetenv utf8 vbell vbell_msg vbellwait version wall
	width windowlist windows wrap writebuf writelock xoff xon zmodem zombie ) );

	for my $name (@commands) {
		install_sub({
			code => sub { shift->send_command($name,@_) },
			as   => $name
		});
	}

	my @rcommands = ( qw( bind meta chdir exec umask) );

	for my $name (@rcommands) {
		install_sub({
			code => sub { shift->send_command($name,@_) },
			as   => "s$name"
		});
	}
}

has session    => (is => 'rw', isa => 'Str' );
has window     => (is => 'rw', isa => 'Str',  default => '0' );
has executable => (is => 'rw', isa => 'Str',  default => sub { which("screen") } );
has create     => (is => 'ro', isa => 'Bool', default => 0 );
has debugging  => (is => 'rw', isa => 'Bool', default => 0 );

sub BUILD {
	my ($self) = @_;
	if ($self->create) {
		if (!$self->session) {
			$self->session("term_gnuscreen.$$" . int(rand(10000)));
		}
		$self->call_screen('-m','-d');
	}
}

sub send_command {
	my ($self,$cmd,@args) = @_;
	$self->call_screen('-X', $cmd, @args) if $cmd;
}

sub call_screen {
	my ($self,@parameter) = @_;
	my @screencmd = ( $self->executable );
	push @screencmd, '-S', $self->session if defined $self->session;
	push @screencmd, '-p', $self->window  if defined $self->window;
	push @screencmd, @parameter;

	if ($self->debugging) {
		print STDERR "Command: " . join(" ",@screencmd) . "\n";
	}

	my ($stdout,$stderr);
	eval { 
		capture { system(@screencmd) } \$stdout, \$stderr;
		1;
	} or do {
		my $err;# = $!;
		$err = $stderr if defined $stderr;
		$err = $stdout if defined $stdout; # '*err*, stdout seems to be actual more helpful
		die "$err";
	};
	return 1;
}

sub hardcopy {
	my ($self,$file) = @_;
	if (!$file) {
		$file = tmpnam();
	}
	$self->send_command('hardcopy',$file);
	return $file;
}

1;

__END__


=head1 NAME

Term::GnuScreen - Control GNU screen via perl

=head1 VERSION

Version 0.03

=head1 SYNOPSIS

Term::GnuScreen provides a simple interface to control a GNU screen
session via its command line interface.

    use Term::GnuScreen;

    my $screen = Term::GnuScreen->new();
    $screen->windowlist;
    $screen->hardcopy('/tmp/my_hardcopy');

=head1 CONSTRUCTION

=over 4

=item session

Sets the name of the screen session to send commands to. If you also
set C<create> to a true value, this will become the new name of your
screen session. See I<-S> option for screen for a further discussion of
this argument.

=item create

If create is set to a true value, a new screen session is created
and detached automatically. If you do not provide a session name,
this module generates one by calling C<"term_gnuscreen" . $$
. int(rand(10000))>. Settings this value after object creation has no
effect at the moment.

The newly created session will not be terminated after programm execution.

=item window

Preselects a window to send a command via the a specific window. Defaults
to 0. See I<-p> option for screen for a further discussion of this
argument.

=item executable

Return or set the screen binary to call. Defaults to the binary found
by File::Which::which.

=back

=head1 METHODS

Term::GnuScreen implements all commands as stated in the texinfo document
shipped with GNU screen. To call a commands, it's send via GNU screens -X
paramter to the first running screen session and its current window. You
can change session and window with the according object methods and
construction paramters. Unless listed here, all remaining arguments are
handled over to screen without further modification.

The five commands bind, meta, chdir, exec and umask are prefixed with a
I<s> ( sbind, smeta, schdir, sexec and sumas ) to distinguish them from
the built-ins with the same name.

=head2 call_screen

This command is the working horse of Term::GnuScreen. It simply builds
the command line to call and execute it.

=head2 send_command

Calls call_screen with the I<-X> and all supplied paramters. Most
functions are implemented by this method.

=head2 hardcopy

Write a hardcopy of the current window to a temporary file and returns
the filename unless the filename is supplied as first argument. If
the supplied filename is not absolute, the file is written relative to
C<hardcopydir>.

=head1 ERROR HANDLING

Simple dies in case screen -X did not return with a return value of
zero. Either $!, STDERR or STDOUT (which seems to be more helpful
most times) are provided as error message for further investigation.

=head1 AUTHOR

Mario Domgoergen, C<< <dom at math.uni-bonn.de> >>

=head1 BUGS AND LIMITATIONS

It seems not to be possible to question a specific screen session
about its state, so this module basically just sends commands to a
screen session without knowing if the command succeeded or was at least
syntactically corrent.

This module needs a lot more testing.

Please report any bugs or feature requests to
C<bug-term-gnuscreen at rt.cpan.org>, or through the web interface
at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Term-GnuScreen>.
I will be notified, and then you'll automatically be notified of progress
on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Term::GnuScreen

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Term-GnuScreen>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Term-GnuScreen>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Term-GnuScreen>

=item * Search CPAN

L<http://search.cpan.org/dist/Term-GnuScreen>

=back


=head1 ACKNOWLEDGEMENTS

L<screen>

=head1 COPYRIGHT & LICENSE

Copyright 2009 Mario Domgoergen, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.
